src/pushjack/apns.py

```diff
@@ -764,6 +778,297 @@ class APNSExpiredToken(namedtuple('APNSExpiredToken', ['token', 'timestamp'])):
+                ``None``.
+            sound (str, optional): Name of the sound file to play for alert.
+                Defaults to ``None``.
+            category (str, optional): Name of category. Defaults to ``None``.
+            content_available (bool, optional): If ``True``, indicate that new
+                content is available. Defaults to ``None``.
+            title (str, optional): Alert title.
+            title_loc_key (str, optional): The key to a title string in the
+                ``Localizable.strings`` file for the current localization.
+            title_loc_args (list, optional): List of string values to appear in
+                place of the format specifiers in `title_loc_key`.
+            action_loc_key (str, optional): Display an alert that includes the
+                ``Close`` and ``View`` buttons. The string is used as a key to
+                get a localized string in the current localization to use for
+                the right button’s title instead of ``"View"``.
+            loc_key (str, optional): A key to an alert-message string in a
+                ``Localizable.strings`` file for the current localization.
+            loc_args (list, optional): List of string values to appear in place
+                of the format specifiers in ``loc_key``.
+            launch_image (str, optional): The filename of an image file in the
+                app bundle; it may include the extension or omit it.
+            mutable_content (bool, optional): if ``True``, triggers Apple
+                Notification Service Extension. Defaults to ``None``.
+            thread_id (str, optional): Identifier for grouping notifications.
+                iOS groups notifications with the same thread identifier
+                together in Notification Center. Defaults to ``None``.
+            extra (dict, optional): Extra data to include with the alert.
+        """
+        assert ids, 'Must provide device IDs.'
+
+        if not isinstance(ids, (list, tuple)):
+            ids = [ids]
+
+        unsendable_devices = []
+        if max_payload_length is None:
+            max_payload_length = self.default_max_payload_length
+
+        message = APNSMessage(message,
+                              max_payload_length=max_payload_length,
+                              **options)
+
+        validate_tokens(ids)
+        validate_message(message, max_size=APNS_HTTP2_MAX_NOTIFICATION_SIZE)
+
+        for device_id in ids:
+
+            try:
+                self._send(
+                    device_id=device_id, message=message,
+                    low_priority=low_priority, expiration=expiration)
+            except Exception:
+                unsendable_devices.append(device_id)
+
+        if not unsendable_devices:
+            return
+
+        if len(unsendable_devices) != len(ids):
+            raise APNSMessagePartiallySentError
+
+        raise APNSUnsendableError("Unable to send messages")
```

Comment:
If you take a look at `APNSConnection.sendall()`, there's an `APNSResponse` that's returned instead of raising an exception. That response object contains useful information for correlating what happened with each message send and why it might have failed. We should do something similar (probably just use `APNSResonse`) here. Otherwise, there's no way to determine what was sent and what wasn't (in case end-user wants to retry sending just the failed ones).